src/main/java/org/opensearch/knn/index/codec/KNN80Codec/KNN80DocValuesProducer.java

```diff
@@ -0,0 +1,154 @@
+    private final SegmentReadState state;
+    private final DocValuesProducer delegate;
+    private final NativeMemoryCacheManager nativeMemoryCacheManager;
+    private final Map<String, String> indexPathMap = new HashMap();
+
+    public KNN80DocValuesProducer(DocValuesProducer delegate, SegmentReadState state) {
+        this.delegate = delegate;
+        this.state = state;
+        this.nativeMemoryCacheManager = NativeMemoryCacheManager.getInstance();
+
+        Directory directory = state.directory;
+        // directory would be CompoundDirectory, we need get directory firstly and then unwrap
+        if (state.directory instanceof KNN80CompoundDirectory) {
+            directory = ((KNN80CompoundDirectory) state.directory).getDir();
+        }
+
+        FSDirectory fsDirectory = null;
+        try {
+            fsDirectory = ((FSDirectory) FilterDirectory.unwrap(directory));
+        } catch (ClassCastException ex) {
+            log.warn("{} can not casting to FSDirectory", directory.toString());
+            return;
+        }
+        assert fsDirectory != null;
+        String directoryPath = fsDirectory.getDirectory().toString();
+        for (FieldInfo field : state.fieldInfos) {
+            if (!field.attributes().containsKey(KNN_FIELD)) {
+                continue;
+            }
+            // Only Native Engine put into indexPathMap
+            KNNEngine knnEngine = getNativeKNNEngine(field);
+            if (knnEngine == null) {
+                continue;
+            }
+            List<String> engineFiles = KNNCodecUtil.getEngineFiles(knnEngine.getExtension(), field.name, state.segmentInfo);
+            Path indexPath = PathUtils.get(directoryPath, engineFiles.get(0));
+            indexPathMap.putIfAbsent(field.getName(), indexPath.toString());
+        }
+    }
+
+    @Override
+    public BinaryDocValues getBinary(FieldInfo field) throws IOException {
+        return delegate.getBinary(field);
+    }
+
+    @Override
+    public NumericDocValues getNumeric(FieldInfo field) throws IOException {
+        return delegate.getNumeric(field);
+    }
+
+    @Override
+    public SortedDocValues getSorted(FieldInfo field) throws IOException {
+        return delegate.getSorted(field);
+    }
+
+    @Override
+    public SortedNumericDocValues getSortedNumeric(FieldInfo field) throws IOException {
+        return delegate.getSortedNumeric(field);
+    }
+
+    @Override
+    public SortedSetDocValues getSortedSet(FieldInfo field) throws IOException {
+        return delegate.getSortedSet(field);
+    }
+
+    @Override
+    public void checkIntegrity() throws IOException {
+        delegate.checkIntegrity();
+    }
+
+    @Override
+    public void close() {
+        try {
+            delegate.close();
+            for (String path : indexPathMap.values()) {
+                nativeMemoryCacheManager.invalidate(path);
+            }
+        } catch (IOException e) {
+            throw new RuntimeException(e);
+        }
```

Comment:
if during the invalidation we get exception from 1 file then it will lead to throwing of exceptions and other graph files will still remain in cache. How we are handling that case?

Reply:
nice catch, let's move invalidate cache ahead

Reply:
FIXED at https://github.com/opensearch-project/k-NN/pull/1946/commits/3b5751c08732d0a147880627fd44ae8b3196fe98